core/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java

```diff
@@ -142,6 +143,10 @@ protected NodeGatewayStartedShards nodeOperation(NodeRequest request) {
                         }
                         Store.tryOpenIndex(shardPath.resolveIndex(), shardId, nodeEnv::shardLock, logger);
                     } catch (Exception exception) {
+                        if (exception instanceof IOException && exception.getCause() instanceof ShardLockObtainFailedException) {
```

Comment:
I can directly throw `ShardLockObtainFailedException`. I wasn't sure why you wrapped that exception in the first place so I left as is.

To your second point, I left the decision on how to treat the `ShardLockObtainFailedException` to `PrimaryShardAllocator` as it has more context available to make the final decision where to allocate the shard. For example, it prioritizes another valid shard copy that has not thrown the exception. Also it allows the shard store action `/_shard_stores` to properly expose the exception as it reuses the same endpoint as the primary shard allocator.